### PR TITLE
SonnyRistrettoPointGadget implementation & refactor

### LINKED ISSUE #12: Refactor to work with SonnyRistrettoPointGadget
Ass discussed with @kev on a call. There are a couple of things we should refactor on this PR in order to merge it.

- [x] Just use `SonnyRistrettoPointGadget` and implement the RistrettoPoint verification inside of the circuit.
- [x] Implement the constrains that we need to apply regarding the previous point like `not_equal`, `ristretto_point_equalty` and `is_not_zero` etc.. -> Done in 958a204.

_Originally posted by @CPerezz in https://github.com/dusk-network/bulletproofs_gadgets/pull/11#issuecomment-584184957_

### PR DESCRIPTION
**Pull request by CPerezz**
Closes #12 by refactoring #11 according to the stated points on the issue #12

### CODE REVIEW

**Comment by CPerezz** [comment]
As seen on the last commit 958a204, the `non_zero` gadget works correctly.
So after building tests for each component, the error has to reside in the way we are computing `8*P`. I think that the points that we are getting from the `double` implementation for `SonnyRistrettoPointGadget` are not being correctly computed for some reason.

**Comment by CPerezz** [comment]
With the latest test on `doubling` without checking ristretto correctness introduced in b1090aa we can see that the `double` implementation for `SonnyRistrettoPointGadget` is failing.
Probably the initial refactor introduced an error somewhere or something similar.

At least, the error is now spotted, so we can now review the `double` impl @decentralisedkev .

**Comment by CPerezz** [comment]
Actually, I noticed that when we commit the Point coordinates using the methods `prover_commit_to_sonnypoint()` or `verifier_commit_to_sonnypoint()`. Without constraining the points to be Ristretto the tests also fail. So it can probably mean that the R1CS that the prover and the verifier have are not the same. we need to dig into it.

**Comment by CPerezz** [comment]
Even the sk_knowledge_gadget works, we are not able to use it with secret points since we are experiencing an error on the verification that most probably relies on the way on which we are generating the `SonnyRistrettoPointGadget` from the prover and verifier perspectives.

Also, point doubling gadget is reliyng now on the addition since it was also having an error that we are still looking for. So #4 cannot be closed yet.

Appart from that, I think we should modularize much better the lib since now all gadgets are mixed on the test file and also on the `ristretto_point.rs` one.

File: Cargo.toml [hunk 1]
```diff
@@ -15,4 +15,7 @@ rand = "0.7.3"
 [dependencies.bulletproofs]
 git = "https://github.com/dusk-network/bulletproofs"
 branch = "develop"
-features = ["yoloproofs"]
\ No newline at end of file
+features = ["yoloproofs"]
+
+[profile.dev]
+opt-level = 3
```
Comment (vlopes11) [inline Cargo.toml]: I don't understand this part. The dev/check build gets too slow; if we need to run tests, we could do something like `$ cargo test --release`?

File: src/gadgets/sk_knowledge/sk_know_gadget.rs [hunk 2]
```diff
@@ -1,22 +1,122 @@
...
-// 1 0 0 1 1 1 0 1 ...... 0 1 1 0
+fn sk_knowledge_proof(
+    pc_gens: &PedersenGens,
+    bp_gens: &BulletproofGens,
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk: &[Scalar],
+) -> Result<(R1CSProof, Vec<CompressedRistretto>), R1CSError> {
+    // Generate transcript
+    let mut transcript = Transcript::new(b"Sk_knowledge");
```
Comment (vlopes11) [inline src/gadgets/sk_knowledge/sk_know_gadget.rs:45]: Just a very minor issue, but generally we have two strategies for the transcript: either we receive it via parameter, as in common dalek's implementations (e.g. zkp crate), or we create a function to return the bp_gens, transcript and pc_gens (my personal favorite)

File: src/gadgets/sk_knowledge/sk_know_gadget.rs [hunk 3]
```diff
@@ -1,22 +1,122 @@
...
+    verifier.verify(proof, pc_gens, bp_gens, &mut rand::thread_rng())
+}
+
+fn sk_knowledge_gadget_roundtrip_helper(
+    basep: SonnyRistrettoPoint,
+    pk: SonnyRistrettoPoint,
+    sk: SonnyScalar,
+) -> Result<(), R1CSError> {
+    let pc_gens = PedersenGens::default();
+    let bp_gens = BulletproofGens::new(8192, 1);
```
Comment (vlopes11) [inline src/gadgets/sk_knowledge/sk_know_gadget.rs:92]: 8192 is strictly required? We are using 4096 for Phoenix

File: src/ristretto_point.rs [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            X: Scalar::from_bytes_mod_order(point.0.X.to_bytes()).into(),
+            Y: Scalar::from_bytes_mod_order(point.0.Y.to_bytes()).into(),
+            Z: Scalar::from_bytes_mod_order(point.0.Z.to_bytes()).into(),
+            T: Scalar::from_bytes_mod_order(point.0.T.to_bytes()).into(),
+        };
+        gadget_p.ristretto_gadget(cs, Some(point));
+        gadget_p
+    }
+
+    pub fn from_LCs(lcs: Vec<LinearCombination>, cs: &mut ConstraintSystem) -> Self {
```
Comment (vlopes11) [inline src/ristretto_point.rs]: Very minor issue, but should we not follow the snake case?

File: src/ristretto_point.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+    /// Adds constrains to validate only points that lie on the prime sub-group and excludes the others
+    /// that lie on smaller order groups with order (2, 4 and 8).
+    /// It also adds constrains that validate only points that satisfy the Sonnycurve equation.
+    pub fn ristretto_gadget(
+        &self,
+        cs: &mut dyn ConstraintSystem,
+        point_assign: Option<SonnyRistrettoPoint>,
+    ) {
+        // XXX: Here we should check that the point relies on the curve.
```
Comment (vlopes11) [inline src/ristretto_point.rs:55]: I'm not sure that's possible on zero-knowledge. We are talking about reassuring the curve consistency outside the circuit?

File: src/util.rs [hunk 6]
```diff
@@ -1,47 +1,86 @@
...
+) {
+    let (inv_var, _, _) = cs
+        .allocate_multiplier(var_assigment.and_then(|q| {
+            Some((
+                Scalar::from_bytes_mod_order(q.inverse().to_bytes()),
+                Scalar::one(),
+            ))
+        }))
+        .unwrap();
+    println!("{:?}", cs.multipliers_len());
```
Comment (vlopes11) [inline src/util.rs]: Little reminder to remove this line

**Comment by vlopes11** [comment]
All looks fine as-is! Just pointed a few minor issues that can be solved on further refactoring

File: Cargo.toml (hunk 1 above)
Comment (CPerezz) [reply Cargo.toml]: Ye, I had the Cfg test option set, will remove it! 
Thanks mate!

File: src/gadgets/sk_knowledge/sk_know_gadget.rs (hunk 3 above)
Comment (CPerezz) [reply src/gadgets/sk_knowledge/sk_know_gadget.rs:92]: Yes, we can't do it with less multiply gates.. Maybe #10 reduces the amount of gates, but atm this is needed..

Comment (CPerezz) [reply src/gadgets/sk_knowledge/sk_know_gadget.rs:92]: Theres aprox 7000 of them

File: src/ristretto_point.rs (hunk 4 above)
Comment (CPerezz) [reply src/ristretto_point.rs]: Well, it was more clear that we were refering to LinearCombinations, but I can just allow snake case on this file if the warnings are a problem.

If you think it's better to change it, I'll do it

File: src/ristretto_point.rs (hunk 5 above)
Comment (CPerezz) [reply src/ristretto_point.rs:55]: Inside.

Indeed we need to check inside the R1CS that the points used on the input are really RistrettoPoints. That's why this gadget exists.

File: src/util.rs (hunk 6 above)
Comment (CPerezz) [reply src/util.rs]: Sure!

File: src/gadgets/sk_knowledge/sk_know_gadget.rs (hunk 2 above)
Comment (CPerezz) [reply src/gadgets/sk_knowledge/sk_know_gadget.rs:45]: Not sure I'm following. 

This function is not exposed, it only pretends to be an example of how to implement a `sk_knowledge_proof`.
You should implement it on the way you want on your own crates imo.

**Pull request merged**